src/hooks/usePyodide.tsx

```diff
@@ -390,20 +391,63 @@ async def _install_requirements_from_dir(dir: str) -> None:
 
     for req in reqs:
         req = req.strip()
+        extras = set()
+
         if req == "" or req.startswith("#"):
             continue
         # If it's a URL, then it must be a wheel file.
         if req.startswith("http://") or req.startswith("https://"):
             pkg_name = re.sub(r"^.+/(.*)-\\d.*$", r"\\1", req)
         else:
             # If we got here, it's a package specification.
+            # https://peps.python.org/pep-0508/#examples
             # Remove any trailing version info: "my-package (>= 1.0.0)" -> "my-package"
-            pkg_name = re.sub(r"([a-zA-Z0-9._-]+)(.*)", r"\\1", req).strip()
+            # or "shiny [theme] == 1.2.0" -> "shiny[theme]"
+            pkg_name = re.sub(r" \\[", "[", req)
+            pkg_name = re.sub(r"([a-zA-Z0-9._,\\[\\]-]+)(.*)", r"\\1", pkg_name).strip()
+
+            match_extras = re.match(r"([^\\[]+)(?:\\[(.*)\\])?", pkg_name)
+            if match_extras and match_extras.group(2):
+                pkg_name = match_extras.group(1)
+                extras.update({e.strip() for e in match_extras.group(2).split(",")})
 
-        if pkg_name in micropip.list():
+        if pkg_name not in micropip.list():
+            print(f"Installing {pkg_name} ...")
+            await micropip.install(pkg_name)
+
+        if len(extras) == 0:
             continue
-        print(f"Installing {req} ...")
-        await micropip.install(req)
+        else:
+            # Because micropip.install() doesn't install extras, we have to
+            # find the package requirements of each extra and install them
+            # if they aren't already installed.
+            dist = importlib.metadata.distribution(pkg_name)
+
+            provided_extras = set(dist.metadata.get_all("Provides-Extra") or [])
+            valid_extras = extras & provided_extras
+            invalid_extras = extras - valid_extras
+            if len(invalid_extras):
+                raise ValueError(
+                    f"Invalid extras for package {pkg_name}: {','.join(invalid_extras)}. "
+                    f"Found in '{req}' in requirements.txt."
+                )
+
+            pkg_reqs = dist.requires or []
+
+            for extra in valid_extras:
+                # Convert requires records  : 'libsass>=0.23.0; extra == "theme"'
+                # into just the package name: 'libsass'
+                extra_reqs = [
+                    r for r in pkg_reqs
+                    if f'extra == "{extra}"' in str(r)
+                    or f"extra == '{extra}'" in str(r)
```

Comment:
> Will they always have this exact formatting, with the spaces around `==`?

I think so, or at least if I've found the right documentation it appears that it will: https://packaging.python.org/en/latest/specifications/core-metadata/#provides-extra-multiple-use.

> It would be helpful if ... there was a comment illustrating what the item looks like

Maybe you were scanning quickly; there's exactly this comment two lines above, before `extra_reqs` is defined.